sources/academy/webscraping/scraping_basics_python/07_extracting_data.md

```diff
@@ -0,0 +1,348 @@
+Sony SACS9 10" Active Subwoofer $158.00
+Sony PS-HX500 Hi-Res USB Turntable $398.00
+...
+```
+
+Let's summarize what stands in our way if we want to have it in our Python program as a number:
+
+- A dollar sign preceeds the number,
+- the number contains decimal commas so it better reads by humans,
+- and some prices start with `From`, which reveals there is a certain complexity in how the shop deals with prices.
+
+## Representing price as an interval
+
+The last bullet point is the most important to figure out before we start coding. We thought we'll be scraping numbers, but in the middle of our effort, we discovered that the price is actually an [interval](https://en.wikipedia.org/wiki/Interval_(mathematics)).
+
+In such a situation, we'd normally go and discuss the problem with those who are about to use the resulting data. For their purposes, is the fact that some prices are just minimum prices important? What would be the most useful representation of the interval in the data for them?
+
+Maybe they'd tell us that we can just remove the `From` prefix and it's fine!
+
+```py
+price_text = product.select_one(".price").contents[-1]
+price = price_text.removeprefix("From ")
+```
+
+In other cases, they'd tell us the data must include this information. And in cases when we just don't know, the safest option is to include in the data everything we have from the input and leave the decision on what's important to later stages.
+
+We can represent interval in various ways. One approach could be having min price and max price. For regular prices, these two numbers would be the same. For prices starting with `From`, the max price would be none, because we don't know it:
+
+```py
+price_text = product.select_one(".price").contents[-1]
+if price_text.startswith("From "):
+    min_price = price_text.removeprefix("From ")
+    max_price = None
+else:
+    min_price = price_text
+    max_price = min_price
+```
+
+We're using Python's built-in string methods:
+
+- `.startswith()`, a method for [checking the beginning of a string](https://docs.python.org/3/library/stdtypes.html#str.startswith).
+- `.removeprefix()`, a method for [removing something from the beginning of a string](https://docs.python.org/3/library/stdtypes.html#str.removeprefix).
+
+The whole program now looks like this:
+
+```py
+import httpx
+from bs4 import BeautifulSoup
+
+url = "https://warehouse-theme-metal.myshopify.com/collections/sales"
+response = httpx.get(url)
+response.raise_for_status()
+
+html_code = response.text
+soup = BeautifulSoup(html_code, "html.parser")
+
+for product in soup.select(".product-item"):
+    title = product.select_one(".product-item__title").text
+
+    price_text = product.select_one(".price").contents[-1]
+    if price_text.startswith("From "):
+        min_price = price_text.removeprefix("From ")
+        max_price = None
+    else:
+        min_price = price_text
+        max_price = min_price
+
+    print(title, min_price, max_price)
```

Comment:
I don't think this is how we would represent it in a real dataset, but perhaps it doesn't matter.

The thing is, this makes it hard to know whether the price is real, or an interval. As the consumer of this dataset you have to make additional comparison `if min_price == max_price` to see whether you have an actual price. Might be better to save real prices to `price` and use `min_price` `max_price` only for intervals.

But again, not a strong opinion.

Reply:
I would also mention why there's an interval (because the product has variants with different prices), and that the real prices can be retrieved from the product detail pages, which we will scrape later (we will, right? 😄)

Reply:
I'd love to make it more plausible as a real-world case, but I'm no data analyst, so it's hard for me to assess how this would be best approached. I came up with an interval as something that can capture the nature of the data, but you're right about the additional comparison. Would something like this make sense?

- If there's exact price, both `price` and `min_price` are set to the number
- If there's just minimal price, `price` is set to `None`, and `min_price` is set to the number
- There's no `max_price`

Once we scrape the product pages, we'd just fill in the missing `price` 🤔

Reply:
Addressed in a39846d1